Add HalfToFloat + FloatToHalf operators to PyTorch

radkris-git opened the pull request:
Summary:
Adding two operators
1. at::float_to_half -> Converts FP32 tensor to FP16 tensor
2. at::half_to_float -> Converts FP16 tensor to FP32 tensor.

These operators internally use the kernel provided by FBGeMM. Both C2 and PT will use the same FBGeMM kernel underneath.

Test Plan: buck test caffe2/test:conversion_ops

Differential Revision: D23824869

supriyar commented on aten/src/ATen/native/HalfFloatOps.cpp:
```diff
@@ -0,0 +1,76 @@
+  const auto input_data = input.data_ptr<at::Half>();
+
+#ifdef USE_FBGEMM
+  #ifdef CPU_CAPABILITY_AVX2
```
Can we combine these two `ifdef` statements?

supriyar commented on aten/src/ATen/native/HalfFloatOps.cpp:
```diff
@@ -0,0 +1,76 @@
+} // namespace
+
+
+Tensor half_to_float(const Tensor& input) {
```
Some documentation on these conversion ops wil be useful. How is this different from the existing `.to(dtype)` in pytorch?

supriyar commented on test/test_half_float_ops.py:
```diff
@@ -0,0 +1,45 @@
+class HalfFloatOpsTest(TestCase):
+
+    def test_half_to_float(self):
+        def get_c2_result(weights):
```
I don't think we need to test against C2 for this. Can we add a reference implementation in PT itself?

dzhulgakov reviewed (changes requested):
Please don't add separate ops. There's existing API `x.to(torch.float16)` or `x.half()`. It probably has a terrible implementation on CPU (vanilla tensor iterator) and we should bind it to fbgemm kernel instead. @ngimel can help you with pointers on how to do that

ngimel commented:
Agree with @dzhulgakov, you should provide a more efficient implementation for x.float() and x.half() that go through .to method. The implementations for the `.to` are in TensorConversions.cpp, you are interested in this overload of .to https://github.com/pytorch/pytorch/blob/c1141b6f68dc7a842f3d947ffb3c9f261940d31d/aten/src/ATen/native/TensorConversions.cpp#L86 and to_impl that it calls.

Commit:
Add HalfToFloat + FloatToHalf operators to PyTorch (#45092)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45092

Adding two operators
1. at::float_to_half -> Converts FP32 tensor to FP16 tensor
2. at::half_to_float -> Converts FP16 tensor to FP32 tensor.

These operators internally use the kernel provided by FBGeMM. Both C2 and PT will use the same FBGeMM kernel underneath.

Test Plan:
buck test //caffe2/test:torch -- .*test_half_tensor.*

Run benchmark locally using

```
buck run //caffe2/benchmarks/operator_benchmark/pt:tensor_to_test
```

AI Bench results are pending. I expect that not to finish as we have large queue with jobs pending for 2+ days.

Benchmark for 512x512 tensor with FbGeMM implementation

```
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : short

# Benchmarking PyTorch: FloatToHalfTensorConversionBenchmark
# Mode: Eager
# Name: FloatToHalfTensorConversionBenchmark_M512_N512_cpu
# Input: M: 512, N: 512, device: cpu
Forward Execution Time (us) : 1246.332

# Benchmarking PyTorch: HalfToFloatTensorConversionBenchmark
# Mode: Eager
# Name: HalfToFloatTensorConversionBenchmark_M512_N512_cpu
# Input: M: 512, N: 512, device: cpu
Forward Execution Time (us) : 1734.304
```

Benchmark for 512x512 tensor trunk with no FbGeMM integration.

```
# ----------------------------------------
# PyTorch/Caffe2 Operator Micro-benchmarks
# ----------------------------------------
# Tag : short

# Benchmarking PyTorch: FloatToHalfTensorConversionBenchmark
# Mode: Eager
# Name: FloatToHalfTensorConversionBenchmark_M512_N512_cpu
# Input: M: 512, N: 512, device: cpu
Forward Execution Time (us) : 169045.724

# Benchmarking PyTorch: HalfToFloatTensorConversionBenchmark
# Mode: Eager
# Name: HalfToFloatTensorConversionBenchmark_M512_N512_cpu
# Input: M: 512, N: 512, device: cpu
Forward Execution Time (us) : 152382.494
```

Reviewed By: ngimel

Differential Revision: D23824869

fbshipit-source-id: 8318588d31222243ea4545833cc7b2963c9c95be
